File: lib/factorix/cli/commands/path.rb
# frozen_string_literal: true

require "json"

module Factorix
  class CLI
    module Commands
      # Display Factorio and Factorix paths
      #
      # This command outputs all paths managed by the runtime environment.
      #
      # @example
      #   $ factorix path
      #   executable_path     /path/to/factorio
      #   user_dir            /path/to/user
      #   ...
      class Path < Base
        # Mapping from path type keys to runtime method names
        PATH_TYPES = {
          "executable_path" => :executable_path,
          "user_dir" => :user_dir,
          "mod_dir" => :mod_dir,
          "save_dir" => :save_dir,
          "script_output_dir" => :script_output_dir,
          "mod_list_path" => :mod_list_path,
          "mod_settings_path" => :mod_settings_path,
          "player_data_path" => :player_data_path,
          "lock_path" => :lock_path,
          "current_log_path" => :current_log_path,
          "previous_log_path" => :previous_log_path,
          "factorix_cache_dir" => :factorix_cache_dir,
          "factorix_config_path" => :factorix_config_path,
          "factorix_log_path" => :factorix_log_path
        }.freeze
        private_constant :PATH_TYPES

        # @!parse
        #   # @return [Runtime::Base]
        #   attr_reader :runtime
        #   # @return [Dry::Logger::Dispatcher]
        #   attr_reader :logger
        include Import[:runtime, :logger]

        desc "Display Factorio and Factorix paths"

        option :json, type: :boolean, default: false, desc: "Output in JSON format"

        # Execute the path command
        #
        # @param json [Boolean] output in JSON format
        # @return [void]
        def call(json:, **)
          logger.debug("Displaying all paths")

          result = PATH_TYPES.transform_values {|method_name| runtime.public_send(method_name).to_s }

          if json
            say JSON.pretty_generate(result)
          else
            output_table(result)
          end
        end

        private def output_table(result)
          key_width = result.keys.map(&:length).max
          result.each do |key, value|
            say "%-#{key_width}s  %s" % [key, value]
          end
        end
      end
    end
  end
end

File: internal/cli/path.go
package cli

import (
	"bytes"
	"encoding/json"
	"fmt"

	"github.com/spf13/cobra"

	"github.com/sakuro/factorix/internal/platform"
)

// pathEntries lists the displayed paths in output order.
var pathEntries = []struct {
	key      string
	resolver func(*platform.Runtime) (string, error)
}{
	{"executable_path", (*platform.Runtime).ExecutablePath},
	{"data_dir", (*platform.Runtime).DataDir},
	{"user_dir", (*platform.Runtime).UserDir},
	{"mod_dir", (*platform.Runtime).MODDir},
	{"save_dir", (*platform.Runtime).SaveDir},
	{"script_output_dir", (*platform.Runtime).ScriptOutputDir},
	{"mod_list_path", (*platform.Runtime).MODListPath},
	{"mod_settings_path", (*platform.Runtime).MODSettingsPath},
	{"player_data_path", (*platform.Runtime).PlayerDataPath},
	{"lock_path", (*platform.Runtime).LockPath},
	{"current_log_path", (*platform.Runtime).CurrentLogPath},
	{"previous_log_path", (*platform.Runtime).PreviousLogPath},
	{"factorix_cache_dir", (*platform.Runtime).FactorixCacheDir},
	{"factorix_config_path", (*platform.Runtime).FactorixConfigPath},
	{"factorix_log_path", (*platform.Runtime).FactorixLogPath},
}

func newPathCommand(c *cli) *cobra.Command {
	var jsonOutput bool

	cmd := &cobra.Command{
		Use:   "path",
		Short: "Display Factorio and Factorix paths",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			application, err := c.App()
			if err != nil {
				return err
			}

			keyWidth := 0
			values := make([]string, len(pathEntries))
			for i, entry := range pathEntries {
				value, err := entry.resolver(application.Runtime)
				if err != nil {
					return err
				}
				values[i] = value
				keyWidth = max(keyWidth, len(entry.key))
			}

			p := c.printer(cmd)
			if jsonOutput {
				var buf bytes.Buffer
				buf.WriteString("{\n")
				for i, entry := range pathEntries {
					key, _ := json.Marshal(entry.key)
					value, _ := json.Marshal(values[i])
					fmt.Fprintf(&buf, "  %s: %s", key, value)
					if i < len(pathEntries)-1 {
						buf.WriteString(",")
					}
					buf.WriteString("\n")
				}
				buf.WriteString("}")
				p.Println(buf.String())
				return nil
			}

			for i, entry := range pathEntries {
				p.Printf("%-*s  %s\n", keyWidth, entry.key, values[i])
			}
			return nil
		},
	}
	cmd.Flags().BoolVar(&jsonOutput, "json", false, "Output in JSON format")
	return cmd
}
